boxsdk/object/cloneable.py

```diff
@@ -0,0 +1,47 @@
+# coding: utf-8
+
+from __future__ import unicode_literals, absolute_import
+
+
+class Cloneable(object):
+    """
+    Cloneable interface to be implemented by endpoint objects that should have ability to be cloned, but with a
+    different session member if desired.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(Cloneable, self).__init__(*args, **kwargs)
+
+    def as_user(self, user):
+        """
+        Returns a new endpoint object with default headers set up to make requests as the specified user.
+        :param user:
+            The user to impersonate when making API requests.
+        :type user:
+            :class:`User`
+        """
+        raise NotImplementedError
```

Comment:
`as_user` and `with_shared_link` will only ever be defined one way:

```python
def as_user(self, user):
    return self.clone(self.session.as_user(user))

def with_shared_link(self, shared_link, shared_link_password=None):
    return self.clone(self.session.with_shared_link(shared_link, shared_link_password=shared_link_password))
```

So I think we could move those definitions to here, and then remove that code from other classes.

Doing that would require adding `session` as a `property` or `abstractproperty` of `Cloneable`.

Alternatively, we could do

```python
def __init__(self, session, **kwargs):
    super(Cloneable, self).__init__(**kwargs)
    self._session = session
```

and have `Client` and `BaseEndpoint` pass `session` to `super` instead of setting `self._session` themselves. And then those method definitions I wrote above, can use `self._session` instead of `self.session`.